### PR TITLE
Implementing rustls traits for Hmac and Hash

### PR DESCRIPTION
**Pull request by nnmkhang**
Added:
- glue for Hmac 
- glue for Hash

### CODE REVIEW

File: src/hash.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
+use rustls::crypto::hash;
+use rust_symcrypt::hash::{Sha256State, Sha384State, HashState, sha256, sha384};
```
Comment (PhilHallin) [inline src/hash.rs:3]: >HashState [](http://example.com/codeflow?start=52&length=9)

Do you still use? #Closed

File: src/hash.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    fn output_len(&self) -> usize {
+        32
+    }
+
+    fn start(&self) -> Box<dyn hash::Context> {
+        Box::new(Sha256Context(Sha256State::new()))
+    }
+
+    fn hash(&self, data: &[u8]) -> hash::Output {
```
Comment (PhilHallin) [inline src/hash.rs:21]: >hash [](http://example.com/codeflow?start=7&length=4)

This should be compute #Closed

File: src/hash.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+impl hash::Context for Sha256Context {
+    fn fork_finish(&self) -> hash::Output {
+        let new_context = self.0.copy();
+        hash::Output::new(&new_context.result()[..])
+    }
+
+    fn fork(&self) -> Box<dyn hash::Context> {
+        Box::new(Sha256Context(*self.0.copy()))
```
Comment (PhilHallin) [inline src/hash.rs]: >*self.0.cop [](http://example.com/codeflow?start=31&length=11)

Is "*" necessary here?

self.0.copy() #Closed

Comment (nnmkhang) [reply src/hash.rs]: Yes * dereferences it from the Box<>

File: src/hash.rs (hunk 1 above)
Comment (nnmkhang) [reply src/hash.rs:3]: Needd for the calls since the methods are part of hte HashState trait

File: src/hash.rs (hunk 2 above)
Comment (nnmkhang) [reply src/hash.rs:21]: from rustls code:

`/// Return the output of this hash function with input `data`.
    fn hash(&self, data: &[u8]) -> Output;`

Comment (PhilHallin) [reply src/hash.rs:21]: Did they change the name of this function in the hash::Hash trait:

    /// Return the output of this hash function with input `data`.
    fn compute(&self, data: &[u8]) -> Output;
}

Before it was compute and not hash

Comment (PhilHallin) [reply src/hash.rs:21]: I see that they changed the function name to hash.

File: src/hmac.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+pub struct HmacSha256;
+pub struct HmacSha256Key(HmacSha256State);
+pub struct HmacSha384;
+pub struct HmacSha384Key(HmacSha384State);
+
+
+impl hmac::Hmac for HmacSha256 {
+    fn with_key(&self, key: &[u8]) -> Box<dyn hmac::Key> {
+        Box::new(HmacSha256Key(HmacSha256State::new(key)))
```
Comment (PhilHallin) [inline src/hmac.rs]: >new(key) [](http://example.com/codeflow?start=48&length=8)

This can fail. Add .unwrap() #Closed

Comment (nnmkhang) [reply src/hmac.rs]: Forgot to add it, i have the unwrap()

**Review by PhilHallin — APPROVED** [review]
:shipit:

**Pull request merged**